src/vdaf/mastic.rs

```diff
@@ -379,32 +391,500 @@ where
+            }),
+            (None, Some(verifier_len)) => Ok(MasticPrepareShare {
+                vidpf_proofs: Seed::decode(bytes)?,
+                szk_query_share_opt: Some(SzkQueryShare::<F, SEED_SIZE>::decode_with_param(
+                    &(false, verifier_len),
+                    bytes,
+                )?),
+            }),
+            (None, None) => Ok(MasticPrepareShare {
+                vidpf_proofs: Seed::<SEED_SIZE>::decode(bytes)?,
+                szk_query_share_opt: None,
+            }),
+            (Some(_), None) => Err(CodecError::UnexpectedValue),
+        }
+    }
+}
+
+/// Mastic prepare message.
+///
+/// Result of preprocessing the broadcast messages of both aggregators during the
+/// preparation phase.
+pub type MasticPrepareMessage<F, const SEED_SIZE: usize> = Option<SzkVerifier<F, SEED_SIZE>>;
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticPrepareMessage<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        match self {
+            Some(verifier) => verifier.encode(bytes),
+            None => Ok(()),
+        }
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        match self {
+            Some(verifier) => verifier.encoded_len(),
+            None => Some(0),
+        }
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> ParameterizedDecode<MasticPrepareState<F, SEED_SIZE>>
+    for MasticPrepareMessage<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        prep_state: &MasticPrepareState<F, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match (&prep_state.szk_query_state, prep_state.verifier_len) {
+            (Some(_), Some(verifier_len)) => Ok(Some(
+                SzkVerifier::<F, SEED_SIZE>::decode_with_param(&(true, verifier_len), bytes)?,
+            )),
+            (None, Some(verifier_len)) => Ok(Some(SzkVerifier::<F, SEED_SIZE>::decode_with_param(
+                &(false, verifier_len),
+                bytes,
+            )?)),
+            (None, None) => Ok(None),
+            (Some(_), None) => Err(CodecError::UnexpectedValue),
+        }
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Aggregator<SEED_SIZE, NONCE_SIZE> for Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    type PrepareState = MasticPrepareState<T::Field, SEED_SIZE>;
+
+    type PrepareShare = MasticPrepareShare<T::Field, SEED_SIZE>;
+
+    type PrepareMessage = MasticPrepareMessage<T::Field, SEED_SIZE>;
+
+    /// Begins the Prepare process with the other Aggregators. The [`Self::PrepareState`] returned
+    /// is passed to [`Self::prepare_next`] to get this aggregator's first-round prepare message.
+    ///
+    /// Implements `Vdaf.prep_init` from [VDAF].
+    ///
+    /// [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-08#section-5.2
+    fn prepare_init(
+        &self,
+        verify_key: &[u8; SEED_SIZE],
+        _agg_id: usize,
```

Comment:
Hmm, this is odd. I guess this is because the aggregator ID is encoded by the `VidfpKey` in the `input_share`, so we don't need it here?

Is the aggregator ID encoded on the wire, as part of `input_share`?
* If so, this shouldn't be happening. We should instead get the aggregator ID from here.
* If not, where do we get the aggregator ID from? Regardless, we should enforce here that `agg_id` matches the value we already have in the `VidpfKey`.

My preference would be to remove the aggregator ID from the `input_share` entirely. Perhaps change the type of `input_share.vidfp_key` to `Seed`.

Reply:
It's encoded on the wire within the vidpfkey within the input share, but not separately. Happy to take it out.

Reply:
Yes, it's being encoded within the VidpfKeys. Changing this.